[WIP] implement BigInt (and BigFloat?) with native Array

rfourquet opened the pull request:
This implements what was on the mind of many. `BigInt` has now two fields, an array (of type `Limb` to keep compatibility with GMP, but ideally would be changed to `UInt`), and a `size` like before (again for compatibility, but also to allow in the future the optimization to store small integers inline, which for now is not convincing (AFAICT), probably because even an `#undef` array field forces the object to be heap allocated).
What was `BigInt` is now called `MPZ`, and there are simple routines to convert back and forth when `ccall`ing into libgmp.

I didn't test much the performances, but for the simple tests of addition given occasionally by @wbhart  or by @skariel [here](https://github.com/JuliaLang/julia/pull/10084#issuecomment-73406923), I get roughly a 3x speedup on my machine, with 2/3 of the number of allocations, but (a bit less than) twice as much total memory allocated.

This is WIP as it's a very mechanical work, so there will be errors and omissions. I think the same transformation can be applied to `BigFloat`.
This is branched off unmerged #13815, so only one new commit here.

rfourquet committed:
merge 2 versions (for GMP 5 & 6) of rand(::UnitRange{BigInt})

As a consequence, using dynamically a version of GMP which
does not match the compile time version is no longer an error
(this patch makes this a warming).
The code is shorter, but we break the MPZ API boundary; this also
allows to put allocating pointer_to_array out of the loop, and get
slightly better performances.

rfourquet committed:
avoid re-allocations in rand(UnitRange(::BigInt)) for some cases

rfourquet committed:
implement BigInt with native Array

rfourquet commented:
Ah yes indeed. I know nothing about multi-threading in Julia, so will need help to make this thread-safe.

yuyichao commented:
The simplest solution (assuming the current one is safe for single thread) AFAIK is to make `_*` an array and use sth like

```jl
const _Z = MPZ[]

function __init__()
    resize!(_Z, Base.Threads.nthreads())
    # Initialize `_Z` elements
end

function fancy_gmp_wrapper()
    @inbounds _Z_for_current_thread = _Z[Base.Threads.thread_id()]
    # Use _Z_for_current_thread
end
```

Note that the current version might not be safe due to issues like https://github.com/JuliaLang/julia/issues/16880 since the finalizers can call julia functions (and possibly GMP ones) from within libgmp.

How much is the cost of the libgmp init function? It might be simpler to just create a new one everytime instead of dealing with threading and entrance issue before porting those to native julia, assuming the performance isn't worse than before.

rfourquet commented:
Ok the solution with `thread_id` seems straightforward enough. But I'm not sure to fully understand the rest of the comment. I think `init` is quite cheap, and for the read-only wrappers (`_U, _V, _W`) it would be fine to create new `MPZ` objects each time instead of those globals. If you are refering to the risk of the `MPZ` finalizers, then as stated in the file comments, the simplest would be to get rid of them, as we need only 3 global objects per thread, which much live for the whole Julia session.

yuyichao commented:
The issue is we need more than 3 per thread. These functions can be called from within libgmp again due to finalizers.

rfourquet commented:
What are "these functions"?

yuyichao commented:
Any functions. In particular, functions that refer to your global variables (`_*`s)

rfourquet commented:
I'm sorry, I cannot follow you and understand the problem. And are you referring specifically to the finalizers attached to `MPZ` objects in `init`?

yuyichao commented:
No. I'm refering to user finalizers that can call any functions.

rfourquet commented:
How those finalizers make this PR unsafe? Does it apply to multi-threaded programs only?

yuyichao commented:
Similar problems is everywhere in the runtime and we need to fix them all. We should just not adding more to fix

rfourquet commented:
Here the point of the globals was to avoid allocating work memory in an MPZ each time a new value is computed by libgmp, I don't see yet a nice solution to the GC-race problem (maybe manually allocating/freeing would not be so expensive after all). Or using a mutex?

timholy commented:
Cross-linking to mention of an alternative: https://groups.google.com/d/msg/julia-users/EzMFkgKb4oY/l5gxrg9CCAAJ. Because that's a finite-precision implementation, that doesn't supersede this effort. I also have no plans to flesh it out, but I'd be happy if others ran with that ball.

JeffBezanson commented on base/gmp.jl:
```diff
@@ -30,43 +30,85 @@ const GMP_BITS_PER_LIMB = gmp_bits_per_limb()
+
+# to wrap an existing BigInt x as an MPZ, we call a global MPZ object _W on x: _W(x)
+# the syntax MPZ(x) would work, but would allocate a new MPZ on the heap each time
+function (z::MPZ)(x::BigInt)
```
It's a bit strange to make a number callable. This code also isn't memory safe, since it saves `pointer(x.d)` longer than `x.d` might exist.

rfourquet replied:
Ah yes, I had meant to remove the `Integer` inheritance from `MPZ`! more seriously, I chose this syntax for its lightness, `getindex` would do it too. I'm fine with using a normal function, but I believe this would uncessarily overload the many places where it is used.

JeffBezanson commented on base/gmp.jl:
```diff
@@ -30,43 +30,85 @@ const GMP_BITS_PER_LIMB = gmp_bits_per_limb()
+# _X, _Y, _Z are read-write MPZ objects
+const _X = MPZ()
+const _Y = MPZ()
+const _Z = MPZ()
```
Sadly not thread-safe. Of course Base is not fully thread-safe yet in other places, but we should start heading in that direction.

rfourquet replied:
Are there other similar cases in Base which have taken the lead on thread-safety, whose example I could follow?

yuyichao replied:
Don't use a global variable.

JeffBezanson commented:
Oops, I see @yuyichao got to these before me.

If we can sort out the few details here I think this is great. Basically a "free" huge speed boost.

simonbyrne commented:
It's a pity we can't use [GMP's custom allocation interface](https://gmplib.org/manual/Custom-Allocation.html) for this (or at least, I can't see how we could).

MPFR should be much easier, as it doesn't require these.

rfourquet commented:
Is there any chance that the GC would (now or in the future) run in a different (identifiable) thread, which would allow to have the same solution, with an array indexed by `thread_id` (outlined by @yuyichao above), work both for multi-threaded programs and the GC?

yuyichao commented:
Not in near future

yuyichao commented:
Sorry... wrong button....

The point is, AFAIK, in the long term (though much shorter term then we can schedule all finalizers on a different thread) we'll remove many of the libgmp function calls and that will not need this wrapper for most cheap functions where the performance difference is measurable. All the solutions for using a global variable here are much more complicated and very hard to make safe and efficient and since they are not needed in the long term, it's better to just not use that at all.

rfourquet commented:
@timholy I would love to have "fixed-finite-arbitrary" precision efficient big integers available. FYI, I had tried to implement this PR using `NTuple`, but it's not more efficient than using arrays as of now. Also the immutabilty makes it more complicated to work with gmplib, and I believe that `Base.BigInt` should have a mutable buffer anyway, to allow packages to develop efficient algorithms which are allowed to mutate the data when they "own" the `BigInt`, without having to use an alternative competitive type.

rfourquet commented:
Yes one of the point of this PR is to make it easier to move away from libgmp in base, and making it easier to incrementally rewrite some functions in pure Julia.
For the multi-thread problem, your solution with an array of globals seems fine. For the GC race-problem, I was suggested that I could simply turn-off the GC when accessing those globals.

rfourquet commented:
But is it really that the only solution/official guideline will be for a long time: "you cannot use global data in any form, because it's not GC-safe". Using "work buffers" in functions seems pretty standard practice (wrapped in let blocks in Julia or static local variables in C functions)... some prominent warnings should be put in the docs then.

yuyichao commented:
> Using "work buffers" in functions seems pretty standard practice

And that's why many libc functions has a (possibly non-standard) `_r` counterpart that should be used instead.

rfourquet committed:
add thread/gc safety

Yu Yichao indicated the thread-safety issues and the skeleton of
a solution, and explained the need to keep references to BigInt
before ccall, and suggested to add a field at the end of MPZ.
He also pointed at the GC un-safety of the previous solution
due to the fact that finalizers can call arbitrary code. Scott
Jones suggested to disable the GC to address this specific problem.

timholy commented:
>  it's not more efficient than using arrays as of now.

I haven't followed this PR, but if that's the case, awesome! But perhaps we are talking about different things---if you are willing to live with fixed precision, fundamentally it seems impossible for an `Array` implementation to compete with tuples. I should clarify that when I say "tuples," I mean from-the-ground-up-implemented-as-`NTuple{N,UInt32}`, i.e., this is "taking a library written in a slow language (C :smile:) and rewriting it in julia to make it much faster." If each `BigInt` corresponds to an `Array`, then creating `BigInt`s means allocating `Array`s, one per `BigInt`. Thanks to Julia's wizardry, tuples (of machine types) don't typically trigger allocations (esp. with heavy use of `@inline`), and you can pack them tightly into a container. By itself, the lack of allocation means that tuples should win handily. In case you didn't look at that post carefully, I got a 40x speedup on `+` compared to our current `BigInt` implementation even when no gc was triggered---it would have been an even bigger improvement if one included the effects of gc.

If you _can't_ live with fixed precision, then `Array`s are clearly the way to go, because there is no type-stable way to implement this with tuples.

rfourquet committed:
use macro instead of do block for providing global variables

rfourquet commented:
@timholy great! What I tried which is not more efficient than using arrays is of course `NTuple{N, UInt}` where `N` is not part of the `BigInt` type, so this variable-length tuples as a field forces heap allocation (type instability). So I share your view that for truly arbitrary-lenght (at run-time), the array is the way to go. Again, I really hope to see what you prototyped available at least in a package. Then it may make sense to design the routines in gmp.jl so that they work on those tuple-based types too.

rfourquet commented:
I updated with an hopefully safer version. 
For each thread, there is a set of variables (wrapped up in a structure) `U, V, W, X, Y, Z`. The GC is simply disabled before accessing those variables (to address the problem that those routines could be called from a finalizer. This adds some little overhead). And `MPZ` has a trailing field to contain a reference to the wrapped `BigInt`. 

Here are some efficiency observations (with rough timings, using `srand(1); @time factor(big(2)^125-1)` as my benchmark) :
1. this new version is 30% slower than the unsafe version (but still more than 60% faster than master)
2. I included an intermediate commit which handles the "bookkeeping" with a do-block instead of a macro: this is 15% slower than with the macro
3. To make it a bit cleaner, I tried the `cconvert` solution to keep a reference to the wrapped `BigInt` (suggested [here](https://github.com/JuliaLang/julia/commit/253f15a46c6bbab6f5c7f4b3ff4a7eea5bee879c#commitcomment-17933487)). I have probably not fully understood what would be the best implementation of this suggestion, but it leads to 40% slowdown.
4. The idea "Don't use a global variable" (https://github.com/JuliaLang/julia/pull/17015#discussion_r67809235) doesn't work: it becomes slower (by about 70%) than master. Although the speed benefit of the PR is not its only point (the other being starting being less dependant on gmp), such a speed regression is not acceptable I think.

PS: @yuyichao already [commented](https://github.com/JuliaLang/julia/commit/54377059e1628c15bc2e5152fa5d1aa4741e29db#commitcomment-17968746) on this update, saying that disabling the GC is not correct, and pointing at #16204, so this PR may have to wait on that one.

yuyichao commented:
For not using a global variable, it seems that you don't actually need to call the gmp init and clear function. Given that this already relies on the gmp internal to not do anything funny about the input and output arguments and that I think the long term goal is to implement most of these in julia anyway, I think the right solution to handle 

1. Have one global `MPZ` initialized at init time (this is mainly to avoid copying code from GMP)
2. When you need an output argument, copy over the statically initialized global MPZ, pass it as argument and use `unsafe_wrap(..., true)` to convert the pointer to a julia array. (This relies on the custom malloc we register to gmp to behave the same as the system one (since we use that to free the pointer))

rfourquet commented:
I don't exactly understand: if you `unsafe_wrap` the global `MPZ` data into the new `BigIng` object, then the global `MPZ` needs to be re-initialized (and a buffer reallocated) for the subsequent call, right? So I don't see what your proposition addresses compared the this branch. Also, in my experiment where I avoid using globals, I initialized a temporary `MPZ` to serve as output argument before each `ccall`; it was then even slower to use `unsafe_wrap(..., true)` when creating the `BigInt` rather than allocating a new array field for it and copying over from the `MPZ` result (at least for `BigInt`'s with 2/3 limbs).

yuyichao commented:
> if you unsafe_wrap the global MPZ data into the new BigIng object

The global one only serves as a statically initialized object, it is never modified.

rfourquet commented:
I'm afraid I still can't make sense of your idea. It the pointer of this `MPZ` is passed to `unsafe_wrap(..., true)`, surely the `MPZ` object cannot be used unchanged (with the same pointer, now owned by a Julia array) for the subsquent call. And I still don't see what this addresses compared to this branch

yuyichao commented:
> copy over the statically initialized global MPZ,

rfourquet commented:
@yuyichao, I had not missed this part, but this doesn't help me understand the points I raised in the last comment. You mean copy the pointer of an array to the `MPZ` object, or the data that the pointer points at? As you said "it's never modified", I deduce that it's the second option. Anyway, which data needs to copied? Usually (and maybe always?), an output argument to a libgmp routine is not read from (i.e. it's a write-only argument). Please be less terse in your comments if they are meant to be understood by me.

rfourquet commented:
Small update on the timings: adding the thread-safety and the safety at the level of `ccall` adds only about 5/10 % overhead compared to the unsafe version. But enabling/disabling the GC for "GC-safety" is 25% slower than the thread-safe version. I hope to shock no-one with the following proposition, but wouldn't it be enough to forbid the use of GMP in finalizers (I believe its use is extemely marginal) until a more efficient general solution to this problem exists ? This would be the occasion to also document that the global RNG cannot be used either, and all other as-of-now GC-unsafe objects/functions.

yuyichao commented:
> You mean copy the pointer of an array to the MPZ object, or the data that the pointer points at?

I mean copying the pointer. And I see where the confusion comes from now, despite being as clear as I could possibly be. In the master version of GMP, `__gmpz_init` initialize the pointer statically, which doesn't seems to be the case on 6.1.0 yet.

> But enabling/disabling the GC for "GC-safety" is 25% slower than the thread-safe version. 

That's why it's a bad idea.

> forbid the use of GMP in finalizers

How do you tell? Mark every function/method/specialization that might call GMP to be not allowed to be used in finalizer?

In any case, see [here](https://github.com/JuliaLang/julia/tree/yyc/bigint) for all what I meant. (apart from the added finalizer and init function, which should not be necessary on newer GMP). The branch is as fast as master and slightly slower than one of my other branch which optimizes `finalizer` call. I think the code there (or similar ones) should be used in the long term for the functions that we still need GMP for and this would be good to go when the simple ones are implemented (+,-, *, for example)

rfourquet commented:
Ah I'm glad you posted it! In fact it's close to what I had in a previous iteration, i.e. to use separate types for intput (`immutable`) and output (`type`) arguments; I then moved to using globals in the name of efficiency, but I agree that this solution is cleaner for the longer term. Also nice to see that you could make this no slower than master, I guess in part because of your correct incantions of `cconvert/unsafe_convert` (when I tried again 2 days ago to avoid globals, I also didn't switch back to using immutables, which could explain the 70% overhead I had). 

PS: I was far from understanding without your code; it's now less relevant to clarify this, but I don't see "one global MPZ initialized at init time" which "is never modified", nor "copy over the statically initialized global MPZ"

> Mark every function/method/specialization that might call GMP to be not allowed to be used in finalizer?

Yes exactly! but my disclaimer is that I have a very narrow view of what finalizers are used for. But I don't see that it's more urgent to have GC-safety for `BigInt` than for e.g. for the global RNG.
Otherwise, enabling/disabling the GC may be bad, but it's the fastest safe solution that we seem to have now.

yuyichao commented:
> PS: I was far from understanding without your code; it's now less relevant to clarify this, but I don't see "one global MPZ initialized at init time" which "is never modified", nor "copy over the statically initialized global MPZ"

Well, for gmp master (or whatever the word is for mercurial) `__gmpz_init` statically initialize in which case you can have a statically initialized one and just copy it every time. The pointer will be copied but that doesn't matter since it will be `realloc`ed anyway and should never be written to. Even the finalizer won't be necessary for most of the case either since it is only necessary when we call GMP multiple times on the same output.

rfourquet commented:
Do we close this as an experiment which will be superseded by other upcoming improvements, or is it worth completing this to get nicer performance in the meantime?

vtjnash commented:
Seems worthwhile to look into again, but I don't know if this is the right approach or if someone would do it differently now

rfourquet commented:
I implemented this in a package, with small integers stored inline, and replaced a bunch of calls to `mpz_` functions by lower-level `mpn_` functions; for the remaining cases, I guess the right approach will be to replace these thread-local objects (used to convert to the arrays into MPZ object on which we call `mpz_` functions) by task-local ones (until the functionalities get implemented in terms of `mpn_` API).

Overall, the current performance is generally better than `BigInt` (even with non-small-inlined integers), but there is more work to be done to make it always better. I also experimented with using `String` instead of `Vector{Limb}` for storage, and performance was no always better, I wonder how it will compare to `Memory{Limb}`. I will try to register this package soon, as it can already be used as a drop-in replacement for `BigInt`.
